### PR TITLE
Export and Import UI

### PR DESCRIPTION
**Pull request by crspeller**
#### Summary
UI for import and export of playbooks

Export Link (downloads on click):
image
image

Import button (instant select file if you only have one team):
image

Import team selection:
image

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-40840
https://mattermost.atlassian.net/browse/MM-40842

### CODE REVIEW

**Comment by crspeller** [comment]
@Phrynobatrachus These jest test failures don't seem related to what I edited here. Have you seen these before? Could you take a look?

**Comment by Phrynobatrachus** [comment]
I haven't seen this before. `ChecklistItemState` is an enum so it should always be "initialized" right? Some quick searching suggests that it might be a circular dependency issue, I'll keep looking.

File: webapp/src/types/playbook.ts [hunk 1]
```diff
@@ -1,6 +1,8 @@
 // Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
 // See LICENSE.txt for license information.
 
+import {getAPIUrl} from 'src/client';
```
Comment (Phrynobatrachus) [inline webapp/src/types/playbook.ts:4]: I think this is the reason for the Jest failures. 

`src/client` imports `src/types/playbook`

Comment (crspeller) [reply webapp/src/types/playbook.ts]: Thanks! Weird that it only complains in the tests.

Comment (Phrynobatrachus) [reply webapp/src/types/playbook.ts]: Yeah it is a bit odd, there's apparently a decent amount more (23 according to [this Webpack plugin](https://github.com/aackerman/circular-dependency-plugin#basic-usage) I tested with real quick) that don't seem to cause issues.

File: webapp/src/components/backstage/playbook_list.tsx [hunk 2]
```diff
@@ -123,6 +151,41 @@ const PlaybookList = () => {
...
+                        { teams.length <= 1 &&
+                        <ImportButton
+                            onClick={() => {
+                                if (fileInputRef && fileInputRef.current) {
+                                    fileInputRef.current.click();
+                                }
+                            }}
+                            spin={importUploading}
+                        />
+                        }
```
Comment (iomodo) [inline webapp/src/components/backstage/playbook_list.tsx:188]:
```js
teams.length > 1 ? <TeamSelector...   :  <ImportButton..
```
might be easier to digest.

File: webapp/src/client.ts [hunk 3]
```diff
@@ -667,3 +676,9 @@ export const doFetchWithoutResponse = async (url: string, options = {}) => {
         url,
     });
 };
+
+export const playbookExportProps = (playbook: Playbook) => {
+    const href = `${getAPIUrl()}/playbooks/${playbook.id}/export`;
```
Comment (iomodo) [inline webapp/src/client.ts:681]: shouldn't we be consistent using `getAPIUrl()` vs `apiUrl`? I'd remove `getAPIUrl()` altogether for consistency

**Review by iomodo — APPROVED** [review]
Looks great! Just minor nits you can ignore

File: webapp/src/client.ts (hunk 3 above)
Comment (crspeller) [reply webapp/src/client.ts:681]: Yup. This happened when I fixed the curricular dependency.

File: webapp/src/components/backstage/playbook_list.tsx (hunk 2 above)
Comment (crspeller) [reply webapp/src/components/backstage/playbook_list.tsx:188]: Personally I think the ternary is more confusing.

File: webapp/src/components/backstage/playbook_list.tsx [hunk 4]
```diff
@@ -125,6 +153,41 @@ const PlaybookList = () => {
                             onSearch={setSearchTerm}
                             placeholder={formatMessage({defaultMessage: 'Search for a playbook'})}
                         />
+                        <HorizontalSpacer size={12}/>
+                        <input
+                            type='file'
```
Comment (calebroseland) [inline webapp/src/components/backstage/playbook_list.tsx:158]:
```suggestion
                            type='file'
                            accept='*.json,application/JSON'
```

**Review by calebroseland — CHANGES_REQUESTED** [review]
Nice to see the whole flow! Works well, though I have a request or two:
- `accept` attr on the input to make file format more obvious
- after successful import: either refresh playbooks list (mirroring the `duplicate` flow), or preferably—navigate to the imported playbook to see/confirm details (same as what happens after `create -> save`)


Some sort of success message would also be nice, but along with `duplicate`, this would perhaps be better as a follow-up ticket.

File: webapp/src/client.ts [hunk 5]
```diff
@@ -231,6 +231,11 @@ export async function archivePlaybook(playbookId: Playbook['id']) {
     return data;
 }
 
+export async function importFile(file: any, teamId: string) {
+    const data = await doPost(`${apiUrl}/playbooks/import?team_id=${teamId}`, file);
+    return data;
+}
```
[the remark this answers is not in the document]
Comment (crspeller) [reply webapp/src/client.ts:237]: Maybe, but I don't want to break the current pattern in that file. Maybe we can do that as a holistic change?

**Comment by crspeller** [comment]
@calebroseland Nice suggestions! I did the redirect to the view page of the new playbook. Feels nicer!

**Review by calebroseland — APPROVED** [review]
Thanks, @crspeller!

**Review by abhijit-singh — APPROVED** [review]
Awesome stuff @crspeller!

A couple of tiny non-blocking nits:
1. Might be valuable to add `playbook` to the name of the exported file? Maybe something like `customer_onboarding_playbook.json` instead of adding `export`. While looking for the file later I'm personally much more likely to search by the word `playbook` than `export` for the playbook I downloaded. Also might help build some brand recognition if that file is shared.
2. Can the team selector dropdown be positioned just below `import` the button instead of overlapping on it?
image
Thanks!

**Pull request merged**